pkg/internalregistry/controller.go

```diff
@@ -0,0 +1,333 @@
+		return err
+	}
+
+	var errs []error
+	for _, serviceAccount := range serviceAccounts {
+		errs = append(errs, c.syncServiceAccount(ctx, serviceAccount))
+	}
+	return errors.Join(errs...)
+}
+
+func (c *imagePullSecretsController) syncServiceAccount(ctx context.Context, serviceAccount *corev1.ServiceAccount) error {
+	var err error
+	// ensure image pull secret name annotation
+	serviceAccount, err = c.syncGeneratedImagePullSecretName(ctx, serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	// get generated image pull secret for the service account
+	imagePullSecret, err := c.imagePullSecretForServiceAccount(ctx, serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	// update registry authentication in the generated image pull secret
+	err = c.syncImagePullSecret(ctx, imagePullSecret, serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	// ensure image pull secret is referenced in service account
+	err = c.syncGeneratedImagePullSecretRef(ctx, serviceAccount, imagePullSecret.Name)
+	if err != nil {
+		return err
+	}
+
+	// cleanup legacy resources
+	err = c.cleanupLegacyResources(ctx, serviceAccount)
+
+	return err
+}
+
+// imagePullSecretForServiceAccount returns a the generated image pull secret that this controller manages.
+// If the image pull secret does not already exist, it will be created. A newly created image pull secret
+// will not have the registry authentication file data.
+func (c *imagePullSecretsController) imagePullSecretForServiceAccount(ctx context.Context, serviceAccount *corev1.ServiceAccount) (*corev1.Secret, error) {
+	// get the generated image pull secret
+	imagePullSecret, err := c.secretLister.Secrets(serviceAccount.Namespace).Get(serviceAccount.Annotations[InternalRegistryImagePullSecretRefKey])
+	if kerrors.IsNotFound(err) {
+		// if image pull secret is not found, create it
+		imagePullSecret = newImagePullSecretForServiceAccount(serviceAccount)
+		var actual *corev1.Secret
+		actual, err = c.kubeClient.CoreV1().Secrets(serviceAccount.Namespace).Create(ctx, imagePullSecret, metav1.CreateOptions{FieldManager: controllerName})
+		if err != nil {
+			c.recorder.Warningf("SecretCreateFailed", "Failed to create %s: %v", resourcehelper.FormatResourceForCLIWithNamespace(imagePullSecret), err)
+			return nil, err
+		}
+		imagePullSecret = actual
+		c.recorder.Eventf("SecretCreated", "Created %s because it was missing", resourcehelper.FormatResourceForCLIWithNamespace(imagePullSecret))
+	}
+	return imagePullSecret, err
+}
+
+func newImagePullSecretForServiceAccount(serviceAccount *corev1.ServiceAccount) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            serviceAccount.Annotations[InternalRegistryImagePullSecretRefKey],
+			Namespace:       serviceAccount.Namespace,
+			OwnerReferences: []metav1.OwnerReference{asOwnerReference(serviceAccount)},
+		},
+		Type: corev1.SecretTypeDockercfg,
+		Data: map[string][]byte{
+			corev1.DockerConfigKey: []byte("{}"),
```

Comment:
Why not initialize with the bound SA token already embedded so that we never go through a "created but not valid" state?

Reply:
I cannot bind the token to an object (the image pull secret) that does not exist. Alternatively, I could not bind it at all, making it only time-bound.